### PR TITLE
Multiple Protocol Support for Cloning in ForkAndCloneForm

### LINKED ISSUE #5418: Github fork/clone dialog change protocol of remote
**Do you want to request a *feature* or report a *bug*?**
feature
**What is the current behavior?**
image
The dialog only allows cloning using ssh protocol. 
**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**

**What is the expected behavior?**
To be able to switch to other protocols like https
**Environment you encounter the issue:**

- GitExtensions version:2.51.5  Probably master too
- GIT version: NA
- OS version: NA
- .NET version: NA

**Did this work in previous version of GitExtensions (which)?**
No

### PR DESCRIPTION
**Pull request by digitalcoyote**
Fixes #5418

Changes proposed in this pull request:
- Add a string[] property to IHostedRepository to indicate supported Protocols
- Add a string property  to IHostedRepository that allows getting/setting the protocol to be used
- Modified GitHubRepo (IHostedRepository) to implement theinterface changes and alter the url of CloneReadWriteUrl depending on which protocol was selected.
- Added a Protocol Dropdown populated by the IHostedRepository (and a label) to ForkAndCloneForm (only available when the selected repository supports more than 1 protocol) to selecting the protocol to use.
- Added a bool parameter (defaults to true) to UpdateCloneInfo to prevent resetting the protocol selection every time the cloneinfotext was updated.
- Extracted the logic that set the cloneInfoTxt label to a method (SetCloneInfoText) so it can be updated when only the CloneReadWriteURL is changed without hitting all the other logic in UpdateCloneInfo as well.
 
Screenshots before and after (if PR changes UI):
**Before**
image

**After: No Support/MultipleSelected**
image

**After: Supported w/ SSH Selected**
image

**After: Supported w/ HTTPS Selected**
image

What did I do to test the code and ensure quality:
Manually tested both SSH and HTTPS with the GitHub plugin.
If Unit Tests are required, I will need to know where to put them as I could not find any for GitHubRepo or ForkAndCloneForm. 

Has been tested on (remove any that don't apply):
- GIT 2.19+
- Windows 10

### CODE REVIEW

**Comment by digitalcoyote** [comment]
I did not end up having to make any changes in the Clone logic as it was already able to handle the https url

File: GitUI/CommandsDialogs/RepoHosting/ForkAndCloneForm.Designer.cs [hunk 1]
```diff
@@ -418,18 +423,52 @@ private void InitializeComponent()
...
             this._cloneSetupGB.Text = "Clone";
             // 
+            // _cloneProtocolGroupBox
+            // 
+            this._cloneProtocolGroupBox.Controls.Add(this._useHttpsRadioButton);
+            this._cloneProtocolGroupBox.Controls.Add(this._useSshRadioButton);
+            this._cloneProtocolGroupBox.Location = new System.Drawing.Point(10, 97);
+            this._cloneProtocolGroupBox.Name = "_cloneProtocolGroupBox";
+            this._cloneProtocolGroupBox.Size = new System.Drawing.Size(200, 18);
+            this._cloneProtocolGroupBox.TabIndex = 25;
```
Comment (RussKie) [inline GitUI/CommandsDialogs/RepoHosting/ForkAndCloneForm.Designer.cs]: Could you please confirm the tab order is/remains logical. If not, please correct it.

File: GitUI/CommandsDialogs/RepoHosting/ForkAndCloneForm.Designer.cs [hunk 2]
```diff
@@ -572,5 +613,8 @@ private void InitializeComponent()
         private System.Windows.Forms.FlowLayoutPanel flowLayoutPanel1;
         private System.Windows.Forms.Label _cloneInfoText;
         private System.Windows.Forms.TableLayoutPanel tableLayoutPanel5;
+        private System.Windows.Forms.Panel _cloneProtocolGroupBox;
+        private System.Windows.Forms.RadioButton _useHttpsRadioButton;
+        private System.Windows.Forms.RadioButton _useSshRadioButton;
```
Comment (RussKie) [inline GitUI/CommandsDialogs/RepoHosting/ForkAndCloneForm.Designer.cs]: Please remove leading `_` for the new controls, it is quite atypical.
If you wish, can fix other controls' names in a separate commit.

Comment (digitalcoyote) [reply GitUI/CommandsDialogs/RepoHosting/ForkAndCloneForm.Designer.cs]: I thought it was odd, but when I added them initially I saw a few others had it so I added it to mine as well. I'll fix it and get the others in line as well.

File: GitUI/CommandsDialogs/RepoHosting/ForkAndCloneForm.Designer.cs (hunk 1 above)
Comment (digitalcoyote) [reply GitUI/CommandsDialogs/RepoHosting/ForkAndCloneForm.Designer.cs]: Appears to be logical, but I have to admit I hadn't even thought to check
forkandclonetaborder

File: GitUI/CommandsDialogs/RepoHosting/ForkAndCloneForm.Designer.cs (hunk 2 above)
Comment (digitalcoyote) [reply GitUI/CommandsDialogs/RepoHosting/ForkAndCloneForm.Designer.cs]: Added a commit to remove the leading `_` from each control.

**Comment by digitalcoyote** [comment]
I'll rework to allow the Plugin to provide it's own supported Protocols (or none) to allow selection. I think using strings defined by the plugin that are used to populate a dropdown, as suggested by @vbjay, should allow for more flexibility and for me to simplify some of my changes. For now, I'm going to bed.

File: GitUI/CommandsDialogs/RepoHosting/ForkAndCloneForm.Designer.cs (hunk 2 above)
Comment (RussKie) [reply GitUI/CommandsDialogs/RepoHosting/ForkAndCloneForm.Designer.cs]: thank you

File: GitUI/CommandsDialogs/RepoHosting/ForkAndCloneForm.Designer.cs (hunk 1 above)
Comment (RussKie) [reply GitUI/CommandsDialogs/RepoHosting/ForkAndCloneForm.Designer.cs]: 👍

**Comment by RussKie** [comment]
> Instead of radio buttons, use drop-down.

I had a small debate in my head about this too. There are Windows design guidelines which talk about dropdown vs radio buttons:
* https://docs.microsoft.com/en-us/windows/desktop/uxguide/ctrl-radio-buttons
* https://docs.microsoft.com/en-us/windows/desktop/uxguide/ctrl-drop

There is no clear cut suggestion to use on vs another, but in this instance I'm ok with the radio buttons.

**Comment by digitalcoyote** [comment]
My thought about using a dropdown was triggered by the idea that with radio buttons, a plugin using HTTP, File:// or some future option would have to go back and modify ForkAndCloneForm to be able to support them. While I am not sold that it is the most appealing way to display the option, it is the most flexible way that should keep future development from having to edit the UI of ForkAndClone in case a plugin wants to support one of the protocols listed on https://git-scm.com/book/en/v2/Git-on-the-Server-The-Protocols (Git, Local, HTTP, HTTPS, SSH). Given that, I think the dropdown allowing the plugin to specify what is supported is the best route.

**Comment by vbjay** [comment]
Bingo.  You got the idea.  git supports git:// file:// ssh and there are things out there like https://keybase.io/blog/encrypted-git-for-everyone.

**Comment by digitalcoyote** [comment]
I have it rewritten to use a string, and I was about to look into standardizing control names in ForkAndCloneForm.cs but it looks like that will suggest a lot of changes across many files (mostly translation files). Is this OK? It would be a commit with all references to the control (eg. "_cloneSetupGB.Text") in ForkAndCloneForm.cs, ForkAndCloneForm.Designer.cs as well as each translation file for that control's id (for consistency)

image

**Comment by vbjay** [comment]
Don't worry about translation.  That is handled at release time.

**Comment by digitalcoyote** [comment]
I switched to strings, removed leading underscores and I believe I only committed the necessary designer files. 
Does this need Unit tests?

**Comment by drewnoakes** [comment]
> removed leading underscores 

Why? This makes the diff hard to read.

**Comment by digitalcoyote** [comment]
> Why? This makes the diff hard to read.

For consistency, I was asked not to lead the names of the new controls with underscores. There are ~80 instances of controls with names beginning with an Underscore in this solution and ~2300 that do not lead with an underscore. Since some of the controls in ForkAndCloneForm.cs (but not all) used leading underscores, they were removed as they were inconsistent with the rest of the solution. It was done in a separate commit to prevent those changes from obscuring the diff of the other changes.

**Comment by vbjay** [comment]
Using strings instead for easier addition.

**Comment by digitalcoyote** [comment]
Do I need to go back and put any more edits on the comments (here) to make the rework easier to follow?

**Comment by RussKie** [comment]
Thank you

**Pull request merged**